internal/router/auth.go

```diff
@@ -0,0 +1,148 @@
+package router
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/razorpay/trino-gateway/internal/boot"
+	"github.com/razorpay/trino-gateway/internal/provider"
+)
+
+type AuthCache struct {
+	cache map[string]struct {
+		Timestamp time.Time
+		Password  string
+	}
+	mu sync.Mutex
+}
+
+/*
+Calls an authentication api,
+With Params-
+{
+	"email": "abc@xyz.com",
+	"token": "token123"
+}
+api returns-
+If Authenticated - {"ok": true}
+If not authenticated- {"ok": false}
+
+@returns-
+boolean{True or False},error_message
+*/
+func authenticate(username string, password string) (bool, error) {
+
+	payload := struct {
+		Username string `json:"email"`
+		Token    string `json:"token"`
+	}{
+		Username: username,
+		Token:    password,
+	}
+
+	payloadBytes, _ := json.Marshal(payload)
+	req, _ := http.NewRequest("POST", boot.Config.Auth.ValidationURL, bytes.NewReader(payloadBytes))
+	req.Header.Set("X-Auth-Token", boot.Config.Auth.ValidationAuthToken)
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return false, err
+	}
+	defer resp.Body.Close()
+
+	respBody, _ := ioutil.ReadAll(resp.Body)
+
+	var data struct {
+		OK bool `json:"ok"`
+	}
+	jsonParseError := json.Unmarshal([]byte(respBody), &data)
+	if jsonParseError != nil {
+		return false, jsonParseError
+	}
+
+	return data.OK, nil
+}
+
+func (authCache *AuthCache) check(username, password string) bool {
+	authCache.mu.Lock()
+	defer authCache.mu.Unlock()
+
+	entry, found := authCache.cache[username]
+	if !found {
+		return false
+	}
+
+	if time.Since(entry.Timestamp) > 10*time.Minute {
+		// If entry is older than 10 minutes, then delete the record and return false
+		delete(authCache.cache, username)
+		return false
+	}
+
+	return entry.Password == password
+}
+
+func (authCache *AuthCache) update(username, password string) {
+	authCache.mu.Lock()
+	defer authCache.mu.Unlock()
+
+	authCache.cache[username] = struct {
+		Timestamp time.Time
+		Password  string
+	}{
+		Timestamp: time.Now(),
+		Password:  password,
+	}
+}
+
+func WithAuth(ctx *context.Context, h http.Handler) http.Handler {
+	authCache, ok := (*ctx).Value("authCache").(*AuthCache)
+	if !ok {
+		authCache = &AuthCache{
+			cache: make(map[string]struct {
+				Timestamp time.Time
+				Password  string
+			}),
+		}
+		*ctx = context.WithValue(*ctx, "authCache", authCache)
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+		var trinoUser = "X-Trino-User"
+		var trinoPassword = "X-Trino-Password"
+
+		username := r.Header.Get(trinoUser)
+		password := r.Header.Get(trinoPassword)
+
+		if authCache.check(username, password) {
```

Comment:
Refactor, move this to `authenticate method

This method `WithAuth` will only extract values from request headers and call the `authenticate` method,
and use the returned values from it